src/lib/integration_test_lib/stack_driver_log_engine.ml

```diff
@@ -0,0 +1,513 @@
+                 | `Milliseconds of int64 ]
+      -> t
+      -> unit Deferred.Or_error.t =
+   fun ~blocks ~epoch_reached ~timeout t ->
+    if blocks = 0 && epoch_reached = 0 && timeout = `Milliseconds 0L then
+      Deferred.Or_error.return ()
+    else
+      let now = Time.now () in
+      let timeout_safety =
+        (*Don't wait for more than an hour in any case*)
+        Time.add now (Time.Span.of_ms (Int64.to_float 3600000L))
+      in
+      let query_timeout_ms =
+        match timeout with
+        | `Milliseconds x ->
+            Some (Time.add now (Time.Span.of_ms (Int64.to_float x)))
+        | _ ->
+            None
+      in
+      let timed_out (res : Block_produced_query.Result.Aggregated.t) =
+        match timeout with
+        | `Slots x ->
+            res.slots_passed >= x
+        | `Epochs x ->
+            res.epochs_passed >= x
+        | `Snarked_ledgers_generated x ->
+            res.snarked_ledgers_generated >= x
+        | `Milliseconds _ ->
+            Time.( > ) (Time.now ()) (Option.value_exn query_timeout_ms)
+      in
+      let conditions_passed (res : Block_produced_query.Result.t) =
+        res.block_height >= blocks && res.epoch >= epoch_reached
+      in
+      (*TODO: this should be block window duration once the constraint constants are added to runtime config*)
+      let block_window_duration =
+        Genesis_constants.Constraint_constants.compiled
+          .block_window_duration_ms
+      in
+      let open Deferred.Or_error.Let_syntax in
+      let rec go aggregated_res : unit Deferred.Or_error.t =
+        if Time.( > ) (Time.now ()) timeout_safety then
+          Deferred.Or_error.error_string "wait_for took too long to complete"
+        else if timed_out aggregated_res then Deferred.Or_error.ok_unit
+        else
+          let%bind pull_result =
+            Subscription.pull t.subscriptions.blocks_produced
+          in
+          match Block_produced_query.parse pull_result with
+          | Error e ->
+              Deferred.return (Error e)
+          | Ok None ->
+              Deferred.bind
+                (Async.after
+                   (Time.Span.of_ms (Int.to_float block_window_duration)))
+                ~f:(fun () -> go aggregated_res)
+          | Ok (Some res) ->
+              if conditions_passed res then Deferred.Or_error.ok_unit
+              else
+                go (Block_produced_query.Result.aggregate aggregated_res res)
+      in
+      go Block_produced_query.Result.Aggregated.empty
+
+  let wait_for :
+         ?blocks:int
+      -> ?epoch_reached:int
+      -> ?timeout:[ `Slots of int
+                  | `Epochs of int
+                  | `Snarked_ledgers_generated of int
+                  | `Milliseconds of int64 ]
+      -> t
+      -> unit Deferred.t =
+   fun ?(blocks = 0) ?(epoch_reached = 0) ?(timeout = `Milliseconds 300000L) t ->
+    match%bind
+      Deferred.Or_error.try_with_join (fun _ ->
```

Comment:
Ideally this `try_with_join` should be lowered to be `try_with`s in `wait_for'`

Reply:
removed all the `failwith`s and changed the functions to return `Deferred.Or_error.t`s